### DISCUSSION TITLE
How to add a minmal Webhost to a `HostApplicationBuilder`

### DISCUSSION THREAD

**Discussion by MarcusWichelmann** [dotnet/aspnetcore#54418] [Q&A]
Hi,

I'm building a CLI tool and use [`Host.CreateEmptyApplicationBuilder`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.host.createemptyapplicationbuilder) to configure a simple, empty `HostApplicationBuilder` with just some console logging and dependency injection. This works fine. I really don't need much for this application.

Now, for a single CLI command (one of many), I want to host a gRPC service. For this, I need to use ASP.Net Core. But I don't find a way to add a minimal Kestrel instance to my [`HostApplicationBuilder`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.hostapplicationbuilder).

I don't want to use `WebApplication.CreateBuilder` to create a whole new `WebApplicationBuilder` because I'd have to add my dependencies again to that builder which would duplicate and complicate things. And I also cannot use `WebApplicationBuilder` for all cli commands, because this automatically registers some `IHostedService`s that I don't want to run with each command.

**What I'm searching for:** A way to add just the few dependencies I need for a Kestrel web server with gRPC support to an existing empty `HostApplicationBuilder`. I'd prefer a manual & declarative approach where I can explicitly decide myself, which services I want to include, because I don't need most of the things ASP.Net Core brings by default and want so safe some memory. IMHO, the default `WebApplicationBuilder` does way to much implicit magic that is hidden away in some `internal` classes that cannot be partially reused.

Do you have an idea, how this can be archieved?

**Comment by KennethHoff** [comment]
Would `WebApplication.CreateEmptyBuilder()` be helpful?

**Reply by MarcusWichelmann** [reply]
Thanks, `WebApplication.CreateEmptyBuilder()` goes into the right direction. Still I'd prefer to be able to add a webhost to an existing builder.
